bionic/aip/docker_image_builder.py

```diff
@@ -0,0 +1,129 @@
+"""
+Builds a docker image for Google AI Platform execution using the current Python
+environment.
+"""
+from concurrent.futures.thread import ThreadPoolExecutor
+from textwrap import dedent
+
+from bionic.deps.optdep import import_optional_dependency
+
+import os
+import hashlib
+import sys
+import tempfile
+import re
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+_cached_docker_module = None
+_cached_docker_client = None
+
+
+def get_docker_module():
+    global _cached_docker_module
+
+    if _cached_docker_module is None:
+        _cached_docker_module = import_optional_dependency(
+            "docker", raise_on_missing=True
+        )
+
+    return _cached_docker_module
+
+
+def get_docker_client():
+    global _cached_docker_client
+
+    if _cached_docker_client is None:
+        docker = get_docker_module()
+        logger.info("Initializing Docker client ...")
+        _cached_docker_client = docker.from_env()
+
+    return _cached_docker_client
+
+
+def fix_pip_requirements(pip_requirements) -> str:
+    # There may be entries with editable installs that point to git
+    # repositories. Such entries may specify git+git protocol. which is
+    # deprecated. See https://github.com/pypa/pip/issues/7554.
+    #
+    # Additionally, docker may encounter issues trying to clone a repository
+    # using ssh.
+    return re.sub(r"-e ([^@]*)@", "-e git+https://", pip_requirements.replace(":", "/"))
+
+
+def get_pip_requirements() -> str:
+    return fix_pip_requirements(os.popen("pip freeze").read())
+
+
+def get_image_uri(project_name: str, pip_requirements: str) -> str:
+    m = hashlib.sha256()
+    m.update(pip_requirements.encode("utf-8"))
+    m.update(str(sys.version_info).encode("utf-8"))
+
+    image_tag = f"bionic_{m.hexdigest()}"
+
+    return f"gcr.io/{project_name}/bionic:{image_tag}"
+
+
+def build_image(docker_client, pip_requirements, image_uri):
+    with tempfile.TemporaryDirectory() as tmp_dir:
+        with open(f"{tmp_dir}/requirements.txt", "w") as f:
+            f.write(pip_requirements)
+
+        container_image = f"python:{sys.version_info[0]}.{sys.version_info[1]}"
+
+        with open(f"{tmp_dir}/Dockerfile", "w") as f:
+            f.write(
+                dedent(
+                    f"""
+                    FROM {container_image}
+                    COPY requirements.txt requirements.txt
+                    RUN pip install -r requirements.txt
+                """
+                )
+            )
+
+        logger.info(f"Building {image_uri} using {container_image}")
+        docker_client.images.build(path=tmp_dir, tag=f"{image_uri}")
+
+        logger.info(f"Pushing {image_uri}")
+        for line in docker_client.images.push(f"{image_uri}", stream=True, decode=True):
+            logger.debug(line)
+
+        logger.info(f"Uploaded {image_uri}")
+
+
+def build_image_if_not_exists(project_name) -> str:
```

Comment:
nit: `if_not_exists` doesn't sound grammatical. WDYT about `retrieve_or_build_image_uri`?

Reply:
Hmm, what about `build_image_if_missing` ?

Bionic does not retrieve the image, but only verify if it exists.